barretenberg/cpp/src/barretenberg/dsl/acir_format/acir_format.cpp

```diff
@@ -103,19 +104,47 @@ void build_constraints(Builder& builder, acir_format const& constraint_system, b
         create_block_constraints(builder, constraint, has_valid_witness_assignments);
     }
 
+    // These are set and modified whenever we encounter a recursion opcode
+    //
+    // These should not be set by the caller
+    // TODO: Check if this is always the case. ie I won't receive a proof that will set the first
+    // TODO input_aggregation_object to be non-zero.
+    // TODO: if not, we can add input_aggregation_object to the proof too for all recursive proofs
+    // TODO: This might be the case for proof trees where the proofs are created on different machines
+    std::array<uint32_t, RecursionConstraint::AGGREGATION_OBJECT_SIZE> current_input_aggregation_object = {
+        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
+    };
+    std::array<uint32_t, RecursionConstraint::AGGREGATION_OBJECT_SIZE> current_output_aggregation_object = {
+        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0
+    };
+
     // Add recursion constraints
     for (size_t i = 0; i < constraint_system.recursion_constraints.size(); ++i) {
         auto& constraint = constraint_system.recursion_constraints[i];
-        create_recursion_constraints(builder, constraint, has_valid_witness_assignments);
-
-        // make sure the verification key records the public input indices of the final recursion output
-        // (N.B. up to the ACIR description to make sure that the final output aggregation object wires are public
-        // inputs!)
-        if (i == constraint_system.recursion_constraints.size() - 1) {
-            std::vector<uint32_t> proof_output_witness_indices(constraint.output_aggregation_object.begin(),
-                                                               constraint.output_aggregation_object.end());
-            builder.set_recursive_proof(proof_output_witness_indices);
+        current_output_aggregation_object = create_recursion_constraints(builder,
+                                                                         constraint,
+                                                                         current_input_aggregation_object,
+                                                                         constraint.nested_aggregation_object,
+                                                                         has_valid_witness_assignments);
```

Comment:
doing it this way saves a few thousand gates due to no longer needing to do an assert_eq on the expected output aggregation object. I don't know why it was costing us so many gates, normalization explains some of it, but I would not expect it to be that expensive